File: regions/system/library/go/graphql-client/graphql_client_test.go
package graphqlclient_test

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"

	graphqlclient "github.com/k1s0-platform/system-library-go-graphql-client"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestExecute(t *testing.T) {
	c := graphqlclient.NewInMemoryGraphQlClient()
	ctx := context.Background()

	expected := map[string]any{"user": map[string]any{"id": "1", "name": "Alice"}}
	c.SetResponse("GetUser", expected)

	query := graphqlclient.GraphQlQuery{
		Query:         "query GetUser($id: ID!) { user(id: $id) { id name } }",
		Variables:     map[string]any{"id": "1"},
		OperationName: "GetUser",
	}
	resp, err := c.Execute(ctx, query, nil)
	require.NoError(t, err)
	require.NotNil(t, resp.Data)
	assert.Empty(t, resp.Errors)
}

func TestExecute_NotConfigured(t *testing.T) {
	c := graphqlclient.NewInMemoryGraphQlClient()
	ctx := context.Background()

	query := graphqlclient.GraphQlQuery{
		Query:         "query Unknown { unknown }",
		OperationName: "Unknown",
	}
	_, err := c.Execute(ctx, query, nil)
	require.Error(t, err)
}

func TestExecuteMutation(t *testing.T) {
	c := graphqlclient.NewInMemoryGraphQlClient()
	ctx := context.Background()

	expected := map[string]any{"createUser": map[string]any{"id": "2", "name": "Bob"}}
	c.SetResponse("CreateUser", expected)

	mutation := graphqlclient.GraphQlQuery{
		Query:         "mutation CreateUser($name: String!) { createUser(name: $name) { id name } }",
		Variables:     map[string]any{"name": "Bob"},
		OperationName: "CreateUser",
	}
	resp, err := c.ExecuteMutation(ctx, mutation, nil)
	require.NoError(t, err)
	require.NotNil(t, resp.Data)
	assert.Empty(t, resp.Errors)
}

func TestSetResponse_Overwrite(t *testing.T) {
	c := graphqlclient.NewInMemoryGraphQlClient()
	ctx := context.Background()

	c.SetResponse("Op", "first")
	c.SetResponse("Op", "second")

	query := graphqlclient.GraphQlQuery{Query: "query Op { op }", OperationName: "Op"}
	resp, err := c.Execute(ctx, query, nil)
	require.NoError(t, err)
	data := *resp.Data
	assert.Equal(t, "second", data)
}

func TestInMemoryGraphQlClient_Subscribe(t *testing.T) {
	client := graphqlclient.NewInMemoryGraphQlClient()
	events := []any{
		map[string]any{"id": "1", "name": "Alice"},
		map[string]any{"id": "2", "name": "Bob"},
	}
	client.SetSubscriptionEvents("OnUserCreated", events)

	subscription := graphqlclient.GraphQlQuery{
		Query:         "subscription { userCreated { id name } }",
		OperationName: "OnUserCreated",
	}
	ch, err := client.Subscribe(context.Background(), subscription)
	assert.NoError(t, err)

	count := 0
	for resp := range ch {
		assert.NotNil(t, resp.Data)
		count++
	}
	assert.Equal(t, 2, count)
}

// --- ClientError tests ---

func TestClientError_Error(t *testing.T) {
	tests := []struct {
		err      *graphqlclient.ClientError
		expected string
	}{
		{graphqlclient.NewRequestError("timeout"), "request error: timeout"},
		{graphqlclient.NewDeserializationError("bad json"), "deserialization error: bad json"},
		{graphqlclient.NewGraphQlError("field not found"), "graphql error: field not found"},
		{graphqlclient.NewNotFoundError("user 123"), "not found: user 123"},
	}
	for _, tt := range tests {
		assert.Equal(t, tt.expected, tt.err.Error())
	}
}

func TestClientError_Kind(t *testing.T) {
	assert.Equal(t, graphqlclient.ClientErrorRequest, graphqlclient.NewRequestError("x").Kind)
	assert.Equal(t, graphqlclient.ClientErrorDeserialization, graphqlclient.NewDeserializationError("x").Kind)
	assert.Equal(t, graphqlclient.ClientErrorGraphQl, graphqlclient.NewGraphQlError("x").Kind)
	assert.Equal(t, graphqlclient.ClientErrorNotFound, graphqlclient.NewNotFoundError("x").Kind)
}

func TestClientError_ImplementsError(t *testing.T) {
	var err error = graphqlclient.NewRequestError("test")
	assert.Error(t, err)
}

// --- GraphQlHttpClient tests ---

func TestGraphQlHttpClient_Execute(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		assert.Equal(t, http.MethodPost, r.Method)
		assert.Equal(t, "application/json", r.Header.Get("Content-Type"))

		var query graphqlclient.GraphQlQuery
		err := json.NewDecoder(r.Body).Decode(&query)
		require.NoError(t, err)
		assert.Equal(t, "query GetUser { user { id } }", query.Query)

		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"data":{"user":{"id":"1"}}}`))
	}))
	defer server.Close()

	client := graphqlclient.NewGraphQlHttpClient(server.URL, nil)
	query := graphqlclient.GraphQlQuery{
		Query:         "query GetUser { user { id } }",
		OperationName: "GetUser",
	}
	resp, err := client.Execute(context.Background(), query, nil)
	require.NoError(t, err)
	require.NotNil(t, resp.Data)
	assert.Empty(t, resp.Errors)
}

func TestGraphQlHttpClient_ExecuteMutation(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"data":{"createUser":{"id":"2"}}}`))
	}))
	defer server.Close()

	client := graphqlclient.NewGraphQlHttpClient(server.URL, nil)
	mutation := graphqlclient.GraphQlQuery{
		Query:         "mutation CreateUser { createUser { id } }",
		OperationName: "CreateUser",
	}
	resp, err := client.ExecuteMutation(context.Background(), mutation, nil)
	require.NoError(t, err)
	require.NotNil(t, resp.Data)
}

func TestGraphQlHttpClient_CustomHeaders(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		assert.Equal(t, "Bearer token123", r.Header.Get("Authorization"))
		assert.Equal(t, "custom-value", r.Header.Get("X-Custom"))

		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"data":null}`))
	}))
	defer server.Close()

	headers := map[string]string{
		"Authorization": "Bearer token123",
		"X-Custom":      "custom-value",
	}
	client := graphqlclient.NewGraphQlHttpClient(server.URL, headers)
	query := graphqlclient.GraphQlQuery{Query: "{ ping }"}
	_, err := client.Execute(context.Background(), query, nil)
	require.NoError(t, err)
}

func TestGraphQlHttpClient_ServerError(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
		w.WriteHeader(http.StatusInternalServerError)
		_, _ = w.Write([]byte("internal server error"))
	}))
	defer server.Close()

	client := graphqlclient.NewGraphQlHttpClient(server.URL, nil)
	query := graphqlclient.GraphQlQuery{Query: "{ ping }"}
	_, err := client.Execute(context.Background(), query, nil)
	require.Error(t, err)

	var clientErr *graphqlclient.ClientError
	require.ErrorAs(t, err, &clientErr)
	assert.Equal(t, graphqlclient.ClientErrorRequest, clientErr.Kind)
	assert.Contains(t, clientErr.Message, "500")
}

func TestGraphQlHttpClient_InvalidJSON(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`not json`))
	}))
	defer server.Close()

	client := graphqlclient.NewGraphQlHttpClient(server.URL, nil)
	query := graphqlclient.GraphQlQuery{Query: "{ ping }"}
	_, err := client.Execute(context.Background(), query, nil)
	require.Error(t, err)

	var clientErr *graphqlclient.ClientError
	require.ErrorAs(t, err, &clientErr)
	assert.Equal(t, graphqlclient.ClientErrorDeserialization, clientErr.Kind)
}

func TestGraphQlHttpClient_GraphQlErrors(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"data":null,"errors":[{"message":"field not found"}]}`))
	}))
	defer server.Close()

	client := graphqlclient.NewGraphQlHttpClient(server.URL, nil)
	query := graphqlclient.GraphQlQuery{Query: "{ unknown }"}
	resp, err := client.Execute(context.Background(), query, nil)
	require.NoError(t, err)
	assert.Len(t, resp.Errors, 1)
	assert.Equal(t, "field not found", resp.Errors[0].Message)
}

func TestGraphQlHttpClient_Subscribe_NotSupported(t *testing.T) {
	client := graphqlclient.NewGraphQlHttpClient("http://localhost", nil)
	ch, err := client.Subscribe(context.Background(), graphqlclient.GraphQlQuery{Query: "subscription { x }"})
	require.Error(t, err)
	assert.Nil(t, ch)

	var clientErr *graphqlclient.ClientError
	require.ErrorAs(t, err, &clientErr)
	assert.Equal(t, graphqlclient.ClientErrorRequest, clientErr.Kind)
	assert.Contains(t, clientErr.Message, "does not support subscriptions")
}

func TestGraphQlHttpClient_ConnectionRefused(t *testing.T) {
	client := graphqlclient.NewGraphQlHttpClient("http://127.0.0.1:1", nil)
	query := graphqlclient.GraphQlQuery{Query: "{ ping }"}
	_, err := client.Execute(context.Background(), query, nil)
	require.Error(t, err)

	var clientErr *graphqlclient.ClientError
	require.ErrorAs(t, err, &clientErr)
	assert.Equal(t, graphqlclient.ClientErrorRequest, clientErr.Kind)
}

func TestGraphQlHttpClient_ImplementsInterface(t *testing.T) {
	var _ graphqlclient.GraphQlClient = graphqlclient.NewGraphQlHttpClient("http://localhost", nil)
}

File: regions/system/library/dart/graphql_client/test/graphql_client_test.dart
import 'dart:convert';

import 'package:http/http.dart' as http;
import 'package:http/testing.dart' as http_testing;
import 'package:test/test.dart';
import 'package:k1s0_graphql_client/graphql_client.dart';

void main() {
  late InMemoryGraphQlClient client;

  setUp(() {
    client = InMemoryGraphQlClient();
  });

  group('GraphQlQuery', () {
    test('必須フィールドで生成できること', () {
      const q = GraphQlQuery(query: '{ users { id } }');
      expect(q.query, equals('{ users { id } }'));
      expect(q.variables, isNull);
      expect(q.operationName, isNull);
    });

    test('全フィールドで生成できること', () {
      const q = GraphQlQuery(
        query: 'query GetUser(\$id: ID!) { user(id: \$id) { name } }',
        variables: {'id': '1'},
        operationName: 'GetUser',
      );
      expect(q.operationName, equals('GetUser'));
      expect(q.variables, equals({'id': '1'}));
    });
  });

  group('GraphQlResponse', () {
    test('エラーがない場合にhasErrorsがfalseであること', () {
      const resp = GraphQlResponse<String>(data: 'ok');
      expect(resp.hasErrors, isFalse);
    });

    test('エラーがある場合にhasErrorsがtrueであること', () {
      const resp = GraphQlResponse<String>(
        errors: [GraphQlError(message: 'fail')],
      );
      expect(resp.hasErrors, isTrue);
    });

    test('空リストの場合にhasErrorsがfalseであること', () {
      const resp = GraphQlResponse<String>(errors: []);
      expect(resp.hasErrors, isFalse);
    });
  });

  group('ErrorLocation', () {
    test('行番号と列番号を保持すること', () {
      const loc = ErrorLocation(3, 5);
      expect(loc.line, equals(3));
      expect(loc.column, equals(5));
    });
  });

  group('GraphQlError', () {
    test('メッセージで生成できること', () {
      const err = GraphQlError(message: 'Not found');
      expect(err.message, equals('Not found'));
      expect(err.locations, isNull);
      expect(err.path, isNull);
    });

    test('locationとpathを含めて生成できること', () {
      const err = GraphQlError(
        message: 'err',
        locations: [ErrorLocation(1, 2)],
        path: ['user', 0, 'name'],
      );
      expect(err.locations, hasLength(1));
      expect(err.path, hasLength(3));
    });
  });

  group('ClientError', () {
    test('requestバリアントを生成できること', () {
      final err = ClientError.request('connection refused');
      expect(err.kind, equals(ClientErrorKind.request));
      expect(err.message, equals('connection refused'));
      expect(err.toString(), equals('RequestError: connection refused'));
    });

    test('deserializationバリアントを生成できること', () {
      final err = ClientError.deserialization('invalid json');
      expect(err.kind, equals(ClientErrorKind.deserialization));
      expect(
          err.toString(), equals('DeserializationError: invalid json'));
    });

    test('graphQlバリアントを生成できること', () {
      final err = ClientError.graphQl('field not found');
      expect(err.kind, equals(ClientErrorKind.graphQl));
      expect(err.toString(), equals('GraphQlError: field not found'));
    });

    test('notFoundバリアントを生成できること', () {
      final err = ClientError.notFound('user 123');
      expect(err.kind, equals(ClientErrorKind.notFound));
      expect(err.toString(), equals('NotFoundError: user 123'));
    });

    test('Exceptionのサブタイプであること', () {
      final err = ClientError.request('test');
      expect(err, isA<Exception>());
    });
  });

  group('InMemoryGraphQlClient', () {
    test('executeが設定済みレスポンスを返すこと', () async {
      client.setResponse('GetUser', {
        'data': {'id': '1', 'name': 'Alice'},
      });
      final result = await client.execute(
        const GraphQlQuery(
          query: 'query GetUser { user { id name } }',
          operationName: 'GetUser',
        ),
        (json) => json,
      );
      expect(result.hasErrors, isFalse);
      expect(result.data?['name'], equals('Alice'));
    });

    test('未設定のオペレーションでエラーレスポンスを返すこと', () async {
      final result = await client.execute(
        const GraphQlQuery(query: '{ unknown }', operationName: 'Unknown'),
        (json) => json,
      );
      expect(result.hasErrors, isTrue);
      expect(result.data, isNull);
    });

    test('executeMutationが設定済みレスポンスを返すこと', () async {
      client.setResponse('CreateUser', {
        'data': {'id': '2', 'name': 'Bob'},
      });
      final result = await client.executeMutation(
        const GraphQlQuery(
          query: 'mutation CreateUser { createUser { id name } }',
          operationName: 'CreateUser',
        ),
        (json) => json,
      );
      expect(result.hasErrors, isFalse);
      expect(result.data?['id'], equals('2'));
    });

    test('operationNameがない場合にクエリ文字列をキーとして使用すること', () async {
      client.setResponse('{ me { id } }', {
        'data': {'id': '42'},
      });
      final result = await client.execute(
        const GraphQlQuery(query: '{ me { id } }'),
        (json) => json,
      );
      expect(result.data?['id'], equals('42'));
    });

    test('レスポンスのエラーを返すこと', () async {
      client.setResponse('Fail', {
        'errors': [
          {'message': 'Unauthorized'},
        ],
      });
      final result = await client.execute(
        const GraphQlQuery(query: 'query Fail { fail }', operationName: 'Fail'),
        (json) => json,
      );
      expect(result.hasErrors, isTrue);
      expect(result.errors!.first.message, equals('Unauthorized'));
      expect(result.data, isNull);
    });

    test('subscribeが登録済みイベントを送出すること', () async {
      client.setSubscriptionEvents('OnUserCreated', [
        {'id': '1', 'name': 'Alice'},
        {'id': '2', 'name': 'Bob'},
      ]);

      const subscription = GraphQlQuery(
        query: 'subscription { userCreated { id name } }',
        operationName: 'OnUserCreated',
      );

      final results =
          await client.subscribe(subscription, (json) => json).toList();
      expect(results, hasLength(2));
      expect(results[0].data, isNotNull);
      expect(results[1].data, isNotNull);
    });
  });

  group('GraphQlHttpClient', () {
    http_testing.MockClient mockClient(
      Future<http.Response> Function(http.Request) handler,
    ) {
      return http_testing.MockClient(handler);
    }

    test('executeがPOSTリクエストを送信してレスポンスをパースすること', () async {
      final mock = mockClient((request) async {
        expect(request.method, equals('POST'));
        expect(request.headers['Content-Type'], equals('application/json'));
        final body = jsonDecode(request.body) as Map<String, dynamic>;
        expect(body['query'], equals('{ users { id } }'));
        return http.Response(
          jsonEncode({
            'data': {'id': '1', 'name': 'Alice'},
          }),
          200,
        );
      });

      final httpClient = GraphQlHttpClient(
        'http://localhost:8080/graphql',
        httpClient: mock,
      );

      final result = await httpClient.execute(
        const GraphQlQuery(query: '{ users { id } }'),
        (json) => json,
      );

      expect(result.hasErrors, isFalse);
      expect(result.data?['name'], equals('Alice'));
    });

    test('executeがvariablesとoperationNameを送信すること', () async {
      final mock = mockClient((request) async {
        final body = jsonDecode(request.body) as Map<String, dynamic>;
        expect(body['variables'], equals({'id': '1'}));
        expect(body['operationName'], equals('GetUser'));
        return http.Response(
          jsonEncode({
            'data': {'id': '1'},
          }),
          200,
        );
      });

      final httpClient = GraphQlHttpClient(
        'http://localhost:8080/graphql',
        httpClient: mock,
      );

      await httpClient.execute(
        const GraphQlQuery(
          query: 'query GetUser(\$id: ID!) { user(id: \$id) { id } }',
          variables: {'id': '1'},
          operationName: 'GetUser',
        ),
        (json) => json,
      );
    });

    test('executeがカスタムヘッダーを送信すること', () async {
      final mock = mockClient((request) async {
        expect(request.headers['Authorization'], equals('Bearer token'));
        return http.Response(
          jsonEncode({
            'data': {'ok': true},
          }),
          200,
        );
      });

      final httpClient = GraphQlHttpClient(
        'http://localhost:8080/graphql',
        headers: {'Authorization': 'Bearer token'},
        httpClient: mock,
      );

      await httpClient.execute(
        const GraphQlQuery(query: '{ me { id } }'),
        (json) => json,
      );
    });

    test('404レスポンス時にClientError.notFoundを投げること', () async {
      final mock = mockClient((request) async {
        return http.Response('Not Found', 404);
      });

      final httpClient = GraphQlHttpClient(
        'http://localhost:8080/graphql',
        httpClient: mock,
      );

      expect(
        () => httpClient.execute(
          const GraphQlQuery(query: '{ users { id } }'),
          (json) => json,
        ),
        throwsA(isA<ClientError>().having(
          (e) => e.kind,
          'kind',
          ClientErrorKind.notFound,
        )),
      );
    });

    test('500レスポンス時にClientError.requestを投げること', () async {
      final mock = mockClient((request) async {
        return http.Response('Internal Server Error', 500);
      });

      final httpClient = GraphQlHttpClient(
        'http://localhost:8080/graphql',
        httpClient: mock,
      );

      expect(
        () => httpClient.execute(
          const GraphQlQuery(query: '{ users { id } }'),
          (json) => json,
        ),
        throwsA(isA<ClientError>().having(
          (e) => e.kind,
          'kind',
          ClientErrorKind.request,
        )),
      );
    });

    test('不正なJSONレスポンス時にClientError.deserializationを投げること',
        () async {
      final mock = mockClient((request) async {
        return http.Response('not json', 200);
      });

      final httpClient = GraphQlHttpClient(
        'http://localhost:8080/graphql',
        httpClient: mock,
      );

      expect(
        () => httpClient.execute(
          const GraphQlQuery(query: '{ users { id } }'),
          (json) => json,
        ),
        throwsA(isA<ClientError>().having(
          (e) => e.kind,
          'kind',
          ClientErrorKind.deserialization,
        )),
      );
    });

    test('executeがレスポンスのGraphQLエラーを返すこと', () async {
      final mock = mockClient((request) async {
        return http.Response(
          jsonEncode({
            'errors': [
              {
                'message': 'Unauthorized',
                'locations': [
                  {'line': 1, 'column': 3}
                ],
                'path': ['user'],
              }
            ],
          }),
          200,
        );
      });

      final httpClient = GraphQlHttpClient(
        'http://localhost:8080/graphql',
        httpClient: mock,
      );

      final result = await httpClient.execute(
        const GraphQlQuery(query: '{ user { id } }'),
        (json) => json,
      );

      expect(result.hasErrors, isTrue);
      expect(result.errors!.first.message, equals('Unauthorized'));
      expect(result.errors!.first.locations, hasLength(1));
      expect(result.errors!.first.locations!.first.line, equals(1));
      expect(result.errors!.first.path, equals(['user']));
    });

    test('executeMutationがexecuteと同様に動作すること', () async {
      final mock = mockClient((request) async {
        return http.Response(
          jsonEncode({
            'data': {'id': '2'},
          }),
          200,
        );
      });

      final httpClient = GraphQlHttpClient(
        'http://localhost:8080/graphql',
        httpClient: mock,
      );

      final result = await httpClient.executeMutation(
        const GraphQlQuery(
          query: 'mutation { createUser { id } }',
          operationName: 'CreateUser',
        ),
        (json) => json,
      );

      expect(result.hasErrors, isFalse);
      expect(result.data?['id'], equals('2'));
    });

    test('subscribeがClientError.requestを投げること', () {
      final httpClient = GraphQlHttpClient('http://localhost:8080/graphql');

      expect(
        () => httpClient.subscribe(
          const GraphQlQuery(query: 'subscription { onEvent { id } }'),
          (json) => json,
        ),
        throwsA(isA<ClientError>().having(
          (e) => e.kind,
          'kind',
          ClientErrorKind.request,
        )),
      );
    });

    test('dataがnullの場合にClientError.deserializationを投げること',
        () async {
      final mock = mockClient((request) async {
        return http.Response(jsonEncode({}), 200);
      });

      final httpClient = GraphQlHttpClient(
        'http://localhost:8080/graphql',
        httpClient: mock,
      );

      expect(
        () => httpClient.execute(
          const GraphQlQuery(query: '{ users { id } }'),
          (json) => json,
        ),
        throwsA(isA<ClientError>().having(
          (e) => e.kind,
          'kind',
          ClientErrorKind.deserialization,
        )),
      );
    });
  });
}
